### PR TITLE
Feat/disk netvlad

### PR DESCRIPTION
**Pull request by yzhou-bcom**

### CODE REVIEW

**Review by jim-bcom — CHANGES_REQUESTED** [review]

File: interfaces/api/pipeline/IRelocalizationPipeline.h [hunk 1]
```diff
@@ -75,6 +75,19 @@ class XPCF_CLIENTUUID("597d510d-452a-4da2-9c3a-8d4b8d15c584") XPCF_SERVERUUID("2
                                                          float_t & confidence,
                                                          const SolAR::datastructure::Transform3Df& poseCoarse = SolAR::datastructure::Transform3Df::Identity()) = 0;
 
+    /// @brief Request the relocalization pipeline to process a new image to calculate the corresponding pose and visualize the intermediate results
+    /// @param[in] image: the image to process
+    /// @param[out] currPointCloud: the current 3D point cloud used to compute the pose
```
[the remark this answers is not in the document]
Comment (jim-bcom) [reply interfaces/api/pipeline/IRelocalizationPipeline.h:80]: Ok

File: interfaces/datastructure/GlobalDescriptor.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+     * @brief length of the descriptor, i.e. number of elements stored in the global descriptor
+     * @return number of elements in the global descriptor 
+    */
+    size_t length() const;
+
+    /**
+     * @brief pointer to the descriptor buffer
+     * @return data pointer to the descriptor buffer 
+    */
+    unsigned char* data() const;
```
[the remark this answers is not in the document]
Comment (yzhou-bcom) [reply interfaces/datastructure/GlobalDescriptor.h:75]: this is related to https://discuss.pytorch.org/t/using-torch-from-blob-with-const-data/141597

File: interfaces/datastructure/GlobalDescriptor.h [hunk 3]
```diff
@@ -53,27 +54,32 @@ class SOLARFRAMEWORK_API GlobalDescriptor {
...
-     * @brief set data
+     * @brief constructor with input arguments
      * @param[in] type: global descriptor type 
      * @param[in] dtype: data type of global descriptor 
      * @param[in] buffer: input buffer 
      * @param[in] len: length of the descriptor (number of elements)
-     * @return FrameworkReturnCode::_SUCCESS if succeeded, otherwise FrameworkReturnCode::_ERROR_
     */
-    FrameworkReturnCode setData(GlobalDescriptorType type, GlobalDescriptorDataType dtype, unsigned char* buffer, size_t len);
+    GlobalDescriptor(GlobalDescriptorType type, GlobalDescriptorDataType dtype, unsigned char* buffer, size_t len);
```
Comment (jim-bcom) [inline interfaces/datastructure/GlobalDescriptor.h:63]: Make it private and remove default ctor

File: interfaces/datastructure/GlobalDescriptor.h [hunk 4]
```diff
@@ -100,6 +112,14 @@ class SOLARFRAMEWORK_API GlobalDescriptor {
 
 DECLARESERIALIZE(GlobalDescriptor);
 
+/**
+ * @brief builder of global descriptor 
+ * @param[in] type global descriptor type 
+ * @param[in] buffer data pointer 
+ * @return shared pointer to GlobalDescriptor if success, otherwise std::nullopt 
+*/
+std::optional<SRef<GlobalDescriptor>> buildGlobalDescriptor(GlobalDescriptorType type, unsigned char* buffer);
```
Comment (jim-bcom) [inline interfaces/datastructure/GlobalDescriptor.h:121]: public static in GlobalDescriptor?

File: src/datastructure/GlobalDescriptor.cpp [hunk 5]
```diff
@@ -78,6 +86,23 @@ GlobalDescriptorDataType GlobalDescriptor::getDataType() const
     return m_dataType;
 }
 
+bool GlobalDescriptor::isValid()
```
Comment (jim-bcom) [inline src/datastructure/GlobalDescriptor.cpp:89]: Is this needed, if the builder is used and the checks are already done? The goal is that you cannot create an ill-formed object.

File: src/datastructure/GlobalDescriptor.cpp [hunk 6]
```diff
@@ -78,6 +86,23 @@ GlobalDescriptorDataType GlobalDescriptor::getDataType() const
     return m_dataType;
 }
 
+bool GlobalDescriptor::isValid() 
+{
+    if (!m_buffer) {
```
Comment (jim-bcom) [inline src/datastructure/GlobalDescriptor.cpp:91]:
```cpp
return m_buffer && globalDescriptorToLengthType.find(m_type) != globalDescriptorToLengthType.end() && globalDescriptorToLengthType.at(m_type).first == m_length && globalDescriptorToLengthType.at(m_type).second == m_dataType;
```
?

**Review by jim-bcom — CHANGES_REQUESTED** [review]

File: interfaces/datastructure/GlobalDescriptor.h [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+     * @brief check if content is valid
+     * @return boolean
+    */
+    bool isValid(); 
+
+private:
+    /**
+     * @brief default constructor 
+    */
+    GlobalDescriptor() = default;
```
Comment (jim-bcom) [inline interfaces/datastructure/GlobalDescriptor.h:99]: Remove (is not used, right?)

Comment (yzhou-bcom) [reply interfaces/datastructure/GlobalDescriptor.h:99]: After removing it, solarframework does not compile any more (). "No matching function for call to solar::datastructure::GlobalDescriptor::GlobalDescriptor()  (Frame.cpp:226 IMPLEMENTSERIALIZE(Frame))"  it seems that the default constructor is used for boost serialization,

Comment (jim-bcom) [reply interfaces/datastructure/GlobalDescriptor.h:99]: Ah ok then.

File: src/datastructure/GlobalDescriptor.cpp [hunk 8]
```diff
@@ -43,6 +43,20 @@ std::optional<GlobalDescriptor> GlobalDescriptor::build(GlobalDescriptorType typ
...
+    if (!buffer) {
+        LOG_ERROR("Empty input data buffer, cannot build global descriptor");
+        return nullptr;
+    }
+    if (globalDescriptorToLengthType.find(type) == globalDescriptorToLengthType.end()) {
+        LOG_ERROR("Unsupported type, cannot build global descriptor");
+        return nullptr;
+    }
+    auto gd = GlobalDescriptor(type, globalDescriptorToLengthType.at(type).second, buffer, globalDescriptorToLengthType.at(type).first);
+    return xpcf::utils::make_shared<GlobalDescriptor>(gd);
```
Comment (jim-bcom) [inline src/datastructure/GlobalDescriptor.cpp:57]: nit: you can check, but I think you can pass directly ctor args to `make_shared()` i.o. first creating the object, like:
```cpp
return xpcf::utils::make_shared<GlobalDescriptor>(type, globalDescriptorToLengthType.at(type).second, buffer, globalDescriptorToLengthType.at(type).first);
```

**Review by jim-bcom — APPROVED** [review]

File: src/datastructure/GlobalDescriptor.cpp (hunk 8 above)
Comment (yzhou-bcom) [reply src/datastructure/GlobalDescriptor.cpp:57]: At first, I did this way, but I had a compilation error related to xpcf. So I changed to the current implementation, everything builds well...

Comment (jim-bcom) [reply src/datastructure/GlobalDescriptor.cpp:57]: Ah yes maybe due to the fact that the ctor is now private. Anyway, not a problem, approved.

**Comment by yzhou-bcom** [comment]
Merge branch feat/disknetvlad into master

**Pull request merged**